商品出品機能実装

T0m555 opened the pull request:
# What
商品管理機能を実装
商品管理モデル単体テストを作成

# Why
商品管理機能を実装するため
商品管理機能をテストする単体テストを作成するため

T0m555 committed:
商品登録機能実装

T0m555 committed:
商品登録機能単体テスト実装

T0m555 committed:
修正

コメント削除

kuu0123 commented:
レビュー依頼ありがとうございます！

こちらのプルリクエストですが、実際の画面や機能ができているか確認させていただきたいので、以下のキャプチャ動画をご提出いただけますか？
```
・出品に成功する動画
・出品に失敗しエラーが表示される動画
・ログイン状態のユーザーだけが、商品出品ページへ遷移できる動画
・ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
・金額が入力されたと同時に手数料と利益が表示される動画
・テスト結果の画像
```
「Gyazo　GIF　使い方」とgoogleで検索し、ツールによって生成されたURLをこのPRに貼っていただければ幸いです。
その後、再度レビュー依頼フォームにてプルリクエストURLをご提出いただけますでしょうか。
お手数おかけいたしますが、よろしくお願いします。

T0m555 commented:
・出品に成功する動画
https://gyazo.com/6b2f23703e1eb973de284d0eda214125
・出品に失敗しエラーが表示される動画
https://gyazo.com/5422a89b26d57bc2f79d4bd53b5e5d62
・ログイン状態のユーザーだけが、商品出品ページへ遷移できる動画
・ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/f8d4661730dc56a907b58c43c005be66
・金額が入力されたと同時に手数料と利益が表示される動画
https://gyazo.com/632670f4be866798b19f3c96911d74f9
・テスト結果の画像
https://gyazo.com/dfcfaad2a4fa95e86dcae7b8d2b8b4e8

T0m555 committed:
修正：単体テスト追加

半角数字でないと登録できないテストの追加。

togo-mori commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,24 @@
 class ItemsController < ApplicationController
+
```
ログインしているかどうかの判定はコントローラー側でも行いましょう。
deviseのヘルパーメソッドである `authenticate_user!` を使うと
簡単に実装できますのでこちらを使用しましょう。

**理由**
検証ツールを用いるとログアウト状態でも出品ページへのリンクを作成できてしまうため。

togo-mori commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,24 @@
+  end
+
+  def create
+    @item = Item.create(item_params)
+    if @item.save
```
**実装条件は満たしておりますが、修正をおすすめします。**
```ruby
@item = Item.create(item_params)
```
こちらの記述ですが
```ruby
@item = Item.new(item_params)
```
としましょう。

理由
現状の記述ですと、バリデーションに引っかかる場合もcreateメソッドで
保存処理が行われてしまいます。意図せぬエラーを防ぐためにも
11行目ではnewメソッドでインスタンスを生成するだけにとどめ、
12行目の `if @item.save` で保存するようにすることをおすすめします。

togo-mori commented on app/javascript/item.js:
```diff
@@ -0,0 +1,20 @@
+//   console.log(document.getElementById('hello'));
+// });
+
+function post (){
+  console.log("非同期投稿を実装");
+}
```
```js
function post (){
  console.log("非同期投稿を実装");
}
```
こちらの記述は削除しましょう。

**理由**
デバッグ用のコードは開発中に使用するものであるためです。
コードレビューは一旦完成したコードに対してレビューを行います。

togo-mori commented on app/models/item.rb:
```diff
@@ -1,2 +1,15 @@
+
+  with_options presence: true do
+    validates :price, format: { with: /\A[0-9]+\z/ }
+    validates :price, :numericality => { :greater_than_or_equal_to => 300 }
+    validates :price, :numericality => { :less_than_or_equal_to => 9999999 }
```
```ruby
validates :price, :numericality => { :greater_than_or_equal_to => 300 }
validates :price, :numericality => { :less_than_or_equal_to => 9999999 }
```
こちらの記述ですが、ハッシュロケットは古い記法になるため
シンボルを用いた表記に修正しましょう！
例えば8行目の記述は以下のように修正が可能です。
```ruby
validates :price, numericality:  { greater_than_or_equal_to: 300 }
```

togo-mori commented on README.md:
```diff
@@ -25,20 +25,20 @@ Things you may want to cover:
 | condition_id       | integer         | null: false       |
-| shipping_charge_id | integer         | null: false       |
+| shippingCharge_id | integer         | null: false       |
 | prefecture_id      | integer         | null: false       |
-| days_to_ship_id    | integer         | null: false       |
+| daysToShip_id    | integer         | null: false       |
```
```
| shippingCharge_id | integer         | null: false       |
| daysToShip_id    | integer         | null: false       |
```
こちらのカラムですが、修正前の
```
| shipping_charge_id | integer         | null: false       |
| days_to_ship_id    | integer         | null: false       |
```
に戻しましょう。

理由
railsのカラムは英小文字で実装し、単語の間をアンダースコア(_)でつなぐルールであるため。

togo-mori commented on db/migrate/20210305231813_create_items.rb:
```diff
@@ -1,8 +1,18 @@
+      t.text :description,            null: false
+      t.integer :category_id,         null: false
+      t.integer :condition_id,        null: false
+      t.integer :shippingCharge_id,  null: false
+      t.integer :prefecture_id,       null: false
+      t.integer :daysToShip_id,     null: false
```
```ruby
t.integer :shippingCharge_id,  null: false
t.integer :daysToShip_id,     null: false
```
README同様にマイグレーションでもカラム名を修正しましょう。
詳細は下記コメントをご確認ください。
https://github.com/T0m555/furima-32868/pull/3#discussion_r602853948

togo-mori reviewed (commented):
レビュー依頼ありがとうございます！
いくつかコメントしましたので確認・修正お願いいたします。

修正後再度フォームにてレビュー依頼お願いいたします。

T0m555 committed:
修正

ItemコントローラーのCreateアクションにおいてItem.Createから.newに変更
Javascriptの不要部分削除
ハッシュロケット修正
Railsのルールに従い、カラムは_で区切る名称へ戻した。

aoitech commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,31 @@
+    params.require(:item).permit(:image, :name, :description, :category_id, :condition_id, :shipping_charge_id, :prefecture_id, :days_to_ship_id, :price).merge(user_id: current_user.id)
+  end
+
+  def move_to_index
```
```ruby
def move_to_index
    unless authenticate_user!
      redirect_to action: :index
    end
end
```
こちらですが、現在ログインしていない場合にトップページに戻っていないでしょうか？
トレロの要件に `ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移すること`とございますので、before_actionでauthenticate_user!を呼び出す形に修正しましょう。

上記が完了しましたら、こちらのmove_to_indexは削除していただいて問題ございません。
ご不明点がございましたら、コネクトでご質問いただけますと幸いです。
お手数をお掛けいたしますが、よろしくお願いいたします。

aoitech commented on app/views/items/new.html.erb:
```diff
@@ -1,15 +1,14 @@
 <div class="items-sell-contents">
   <header class="items-sell-header">
+  <%# <script src="item.js"></script> %>
```
```ruby
<%# <script src="item.js"></script> %>
```
こちらのコメントアウトですが、不要であれば削除しましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめします。

aoitech commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+    context '商品登録ができないとき' do
+      it '商品名が入力されていないと登録できない' do
+        @item.name = ''
```
下記のテストコードの記述もお願いいたします。

価格↓
```
半角英数混合では登録できないこと
半角英語では登録できないこと
```

画像(image)↓
```
空では登録できないこと
```

aoitech reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認いただけますと幸いです。
修正後、再度レビュー依頼フォームのご提出をお願いいたします。
お手数をお掛け致しますが、よろしくお願いいたします。

T0m555 committed:
修正

未ログイン時の遷移先がindexになっていたため修正。deviseのauthenticate_user!を用いて実装。
javascript内部の不要なコメントを削除。
単体テストの追加。

T0m555 commented:
・テスト結果の画像
https://gyazo.com/abefccefb4a3a9e93ef4fdb4ca8d12db

aoitech commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,87 @@
+        expect(@item.errors.full_messages).to include("Price must be less than or equal to 9999999")
+      end
+      it '価格が全角数字では登録できない'  do
+        @item.price = "１５００"
```
~~```ruby~~
~~@item.price = "１５００"~~
~~```~~
~~こちら細かい部分ではございますが、数値をテストしたいため、""(ダブルクォーテーション)は外しておきましょう。~~

aoitech reviewed (commented):
~~レビュー依頼ありがとうございます。~~
~~修正点確認いたしました。~~
~~テスト画像の添付もありがとうございます。~~

~~追加で1点のみコメントいたしました。非常に細かい部分で恐縮ではございますが、ご確認いただけますと幸いです。~~

~~修正後再度、レビュー依頼フォームのご提出をお願いいたします。~~
~~お手数をお掛け致しますが、よろしくお願いいたします。~~

aoitech commented:
続けてのコメント失礼いたします。
上記の指摘事項は私の誤りでございました。
半角数字と見間違えておりました。

上記は無視していただけますと幸いです。
この度はお手数をお掛けしまして、大変申し訳ございません。

他に指摘箇所がございませんため、LGTMとさせていただきます。実装お疲れさまでございました。

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切るようにしましょう。
引き続き実装頑張ってください。応援しております。
